File: backend/app/models/local.py
"""Models for the Local SEO suite: GBP manager and citation/NAP distribution."""
from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import Field

from app.models.common import AppModel


# ---------------------------------------------------------------------------
# Google Business Profile (GBP)
# ---------------------------------------------------------------------------


class GbpPostCreate(AppModel):
    """Payload for publishing a GBP post / update."""

    summary: str
    topic: str = "update"  # "update" | "offer" | "event"
    cta_url: str | None = None


class GbpPost(AppModel):
    id: str
    org_id: str
    summary: str
    topic: str = "update"
    cta_url: str | None = None
    state: str = "live"
    created_at: datetime


class GbpMetrics(AppModel):
    """GBP search performance metrics (views, searches, customer actions)."""

    org_id: str
    period_days: int = 30
    views_search: int = 0
    views_maps: int = 0
    searches_direct: int = 0
    searches_discovery: int = 0
    actions_website: int = 0
    actions_calls: int = 0
    actions_directions: int = 0


class ReviewSentiment(str, Enum):
    POSITIVE = "positive"
    NEUTRAL = "neutral"
    NEGATIVE = "negative"


class GbpReview(AppModel):
    id: str
    author: str
    rating: int = Field(default=5, ge=1, le=5)
    text: str = ""
    created_at: datetime
    reply: str | None = None


class SuggestedReply(AppModel):
    """A generated reply suggestion for a review, keyed by rating/sentiment."""

    review_id: str
    sentiment: ReviewSentiment = ReviewSentiment.NEUTRAL
    text: str


# ---------------------------------------------------------------------------
# Citations / NAP distribution
# ---------------------------------------------------------------------------


class Directory(str, Enum):
    YELP = "yelp"
    APPLE_MAPS = "apple_maps"
    BING_PLACES = "bing_places"
    FOURSQUARE = "foursquare"


class NapRecord(AppModel):
    """Canonical org-scoped NAP (name/address/phone) record plus hours/site.

    When ``locked`` is true, drift found in a directory is both flagged and
    blocked: the sync engine re-pushes this canonical record over the drift.
    """

    name: str
    address: str
    phone: str
    website: str | None = None
    hours: dict[str, str] = Field(default_factory=dict)
    locked: bool = False


class NapDiff(AppModel):
    """One field-level difference between the canonical NAP and a listing."""

    field: str
    expected: str
    found: str


class ListingDeliveryState(str, Enum):
    IN_SYNC = "in_sync"
    CREATED = "created"
    UPDATED = "updated"
    FAILED = "failed"
    NOT_LISTED = "not_listed"
    DRIFT_FLAGGED = "drift_flagged"
    DRIFT_BLOCKED = "drift_blocked"


class DirectoryListing(AppModel):
    """The NAP data a directory currently holds for the business."""

    directory: Directory
    name: str = ""
    address: str = ""
    phone: str = ""
    website: str | None = None
    hours: dict[str, str] = Field(default_factory=dict)
    last_synced_at: datetime | None = None


class DirectorySyncResult(AppModel):
    directory: Directory
    state: ListingDeliveryState
    diffs: list[NapDiff] = []
    error: str | None = None


class CitationSyncReport(AppModel):
    """Per-directory delivery state for one sync pass."""

    org_id: str
    results: list[DirectorySyncResult] = []
    delivered: int = 0
    failed: int = 0


class DirectoryStatus(AppModel):
    directory: Directory
    listed: bool = False
    drift: bool = False
    diffs: list[NapDiff] = []
    state: ListingDeliveryState = ListingDeliveryState.IN_SYNC


class CitationStatusReport(AppModel):
    org_id: str
    locked: bool = False
    drift_detected: bool = False
    directories: list[DirectoryStatus] = []

File: mcp-server/src/tools/local.ts
/**
 * Local SEO MCP tools.
 *
 * Google Business Profile management (posts, metrics, reviews + replies) and
 * citation/NAP distribution across directories — backed by /api/v1/local/*.
 */

import { z } from "zod";
import { apiClient } from "../apiClient.js";
import { buildQuery, defineTool, type ToolDefinition } from "./define.js";

/** Directories the citation sync engine can push to (models/local.py Directory). */
const DIRECTORIES = ["yelp", "apple_maps", "bing_places", "foursquare"] as const;

// ---------------------------------------------------------------------------
// local_gbp_post — POST /local/gbp/posts
// ---------------------------------------------------------------------------

const localGbpPostSchema = z.object({
  summary: z.string().min(1, "summary is required"),
  topic: z.enum(["update", "offer", "event"]).optional(),
  cta_url: z.string().optional(),
});

const localGbpPostTool = defineTool(
  "local_gbp_post",
  "Publish a Google Business Profile post/update (update, offer, or event) with an " +
    "optional call-to-action URL.",
  localGbpPostSchema,
  {
    type: "object",
    properties: {
      summary: { type: "string", description: "Post body text shown on the profile" },
      topic: { type: "string", enum: ["update", "offer", "event"], description: "Post type (default update)" },
      cta_url: { type: "string", description: "Call-to-action URL attached to the post" },
    },
    required: ["summary"],
    additionalProperties: false,
  },
  (args) => apiClient.post("/local/gbp/posts", args)
);

// ---------------------------------------------------------------------------
// local_gbp_metrics — GET /local/gbp/metrics
// ---------------------------------------------------------------------------

const localGbpMetricsSchema = z.object({
  period_days: z.number().int().min(1).max(365).optional(),
});

const localGbpMetricsTool = defineTool(
  "local_gbp_metrics",
  "Get Google Business Profile performance metrics for the org over a period: search/maps " +
    "views, direct vs discovery searches, and customer actions (website clicks, calls, " +
    "direction requests).",
  localGbpMetricsSchema,
  {
    type: "object",
    properties: {
      period_days: { type: "integer", minimum: 1, maximum: 365, description: "Reporting period in days (default 30)" },
    },
    required: [],
    additionalProperties: false,
  },
  (args) => apiClient.get(`/local/gbp/metrics${buildQuery({ period_days: args.period_days })}`)
);

// ---------------------------------------------------------------------------
// local_gbp_reviews — GET /local/gbp/reviews
// ---------------------------------------------------------------------------

const localGbpReviewsSchema = z.object({});

const localGbpReviewsTool = defineTool(
  "local_gbp_reviews",
  "List the org's Google Business Profile reviews (author, rating, text, existing reply) " +
    "together with generated sentiment-aware suggested replies for each review. Takes no " +
    "arguments.",
  localGbpReviewsSchema,
  {
    type: "object",
    properties: {},
    required: [],
    additionalProperties: false,
  },
  () => apiClient.get("/local/gbp/reviews")
);

// ---------------------------------------------------------------------------
// local_reply_review — POST /local/gbp/reviews/{review_id}/reply
// ---------------------------------------------------------------------------

const localReplyReviewSchema = z.object({
  review_id: z.string().min(1, "review_id is required"),
  text: z.string().optional(),
});

const localReplyReviewTool = defineTool(
  "local_reply_review",
  "Reply to a Google Business Profile review. Omit 'text' to publish the backend's " +
    "generated sentiment-aware suggested reply for that review.",
  localReplyReviewSchema,
  {
    type: "object",
    properties: {
      review_id: { type: "string", description: "ID of the review to reply to (see local_gbp_reviews)" },
      text: { type: "string", description: "Reply text; omitted = use the generated suggested reply" },
    },
    required: ["review_id"],
    additionalProperties: false,
  },
  (args) =>
    apiClient.post(`/local/gbp/reviews/${encodeURIComponent(args.review_id)}/reply`, {
      text: args.text,
    })
);

// ---------------------------------------------------------------------------
// local_citation_sync — POST /local/citations/sync
// ---------------------------------------------------------------------------

const localCitationSyncSchema = z.object({
  directory: z.enum(DIRECTORIES).optional(),
});

const localCitationSyncTool = defineTool(
  "local_citation_sync",
  "Push the org's canonical NAP (name/address/phone) record to business directories " +
    "(Yelp, Apple Maps, Bing Places, Foursquare) — all of them, or a single one. Returns " +
    "per-directory delivery state (created/updated/in_sync/failed) and field-level diffs.",
  localCitationSyncSchema,
  {
    type: "object",
    properties: {
      directory: {
        type: "string",
        enum: [...DIRECTORIES],
        description: "Sync only this directory (omitted = all directories)",
      },
    },
    required: [],
    additionalProperties: false,
  },
  (args) => apiClient.post(`/local/citations/sync${buildQuery({ directory: args.directory })}`, {})
);

// ---------------------------------------------------------------------------
// local_citation_status — GET /local/citations/status
// ---------------------------------------------------------------------------

const localCitationStatusSchema = z.object({});

const localCitationStatusTool = defineTool(
  "local_citation_status",
  "Check the org's citation health across directories: whether each directory lists the " +
    "business, NAP drift detection with field-level diffs, and lock state. Takes no arguments.",
  localCitationStatusSchema,
  {
    type: "object",
    properties: {},
    required: [],
    additionalProperties: false,
  },
  () => apiClient.get("/local/citations/status")
);

// ---------------------------------------------------------------------------

/** All Local SEO tools exposed by this MCP server. */
export const localTools: ToolDefinition[] = [
  localGbpPostTool,
  localGbpMetricsTool,
  localGbpReviewsTool,
  localReplyReviewTool,
  localCitationSyncTool,
  localCitationStatusTool,
];
